corehq/apps/domain/tests/test_login_and_domain_required.py

```diff
@@ -0,0 +1,149 @@
+        self.addCleanup(self.domain_patcher.stop)
+
+    @staticmethod
+    def _create_domain():
+        domain = mock.Mock(Domain)
+        domain.is_active = True
+        domain.is_snapshot = False
+        return domain
+
+    @staticmethod
+    def _create_user():
+        user = mock.Mock()
+        user.is_superuser = False
+        user.is_member_of = mock.Mock(return_value=False)
+        return user
+
+    def _create_request(self, url, user):
+        request = self.factory.get(url)
+        request.user = request.couch_user = user
+        request.path = 'access_required.html'
+        request.domain = self.domain_name
+
+        return request
+
+    def process_request(self):
+        return process_request(self.request, self.domain_name)
+
+    def test_invalid_domain__displays_404(self):
+        self.get_domain_by_name.return_value = None
+
+        with self.assertRaises(Http404):
+            self.process_request()
+
+    # Reverse is stubbed here because actual reverse lookups take ~ 1 second
+    @mock.patch('corehq.apps.domain.decorators.reverse')
+    @mock.patch('corehq.apps.domain.decorators.messages', mock.Mock())
+    def test_inactive_domain__redirects_to_domain_selection(self, reverse):
+        self.domain.is_active = False
+        reverse.return_value = '/domain/select/'
+
+        response = self.process_request()
+        self.assertEqual(response.status_code, status.HTTP_302_FOUND)
+        reverse.assert_called_with('domain_select')
+
+    def test_two_factor_required_but_missing__displays_alert(self):
+        self.domain.two_factor_auth = True
+        self.user.is_member_of.return_value = True
+        self.user.two_factor_disabled = False
+        self.user.is_verified = mock.Mock(return_value=False)
+
+        response = self.process_request()
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
+
+    def test_snapshot_domain_without_previewer__returns_not_found(self):
+        self.domain.is_snapshot = True
+        self.user.is_previewer = mock.Mock(return_value=False)
+        with self.assertRaises(Http404):
+            self.process_request()
+
+    def test_snapshot_domain_with_previewer__allows_access(self):
+        self.domain.is_snapshot = True
+        self.user.is_previewer = mock.Mock(return_value=True)
+        response = self.process_request()
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_user_belongs_to_domain__returns_200(self):
+        self.user.is_member_of.return_value = True
+        response = self.process_request()
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    @mock.patch('corehq.apps.domain.decorators.has_privilege', mock.Mock(return_value=False))
+    @mock.patch('corehq.apps.domain.views.accounting.SubscriptionUpgradeRequiredView')
+    def test_domain_user_does_not_have_project_access__renders_subscription_upgrade(self, upgrade_required_view_class):  # noqa: E501
```

Comment:
nit: it's not that the domain _user_ doesn't have project access, but that the domain itself doesn't have project access.

Reply:
I don't think that's a nit -- my test name was misleading. Perhaps the code itself should also be updated? The function `_can_access_project_page(request)` led me to believe it was about the user's access. Maybe `project_is_accessible` would better convey what's going on?

Reply:
yeah, I like that much better - as-is it _does_ sound like user access